## Preparation

In [1]:
import numpy as np
import pandas as pd

In [2]:
games = pd.read_csv('appstore_games.csv')

In [3]:
games.head()

URL         ID  \
0   https://apps.apple.com/us/app/sudoku/id284921427  284921427   
1  https://apps.apple.com/us/app/reversi/id284926400  284926400   
2  https://apps.apple.com/us/app/morocco/id284946595  284946595   
3  https://apps.apple.com/us/app/sudoku-free/id28...  285755462   
4  https://apps.apple.com/us/app/senet-deluxe/id2...  285831220   

            Name Subtitle                                           Icon URL  \
0         Sudoku      NaN  https://is2-ssl.mzstatic.com/image/thumb/Purpl...   
1        Reversi      NaN  https://is4-ssl.mzstatic.com/image/thumb/Purpl...   
2        Morocco      NaN  https://is5-ssl.mzstatic.com/image/thumb/Purpl...   
3  Sudoku (Free)      NaN  https://is3-ssl.mzstatic.com/image/thumb/Purpl...   
4   Senet Deluxe      NaN  https://is1-ssl.mzstatic.com/image/thumb/Purpl...   

   Average User Rating  User Rating Count  Price In-app Purchases  \
0                  4.0             3553.0   2.99              NaN   
1                  3.5              284.0   1.99              NaN   
2                  3.0             8376.0   0.00              NaN   
3                  3.5           190394.0   0.00              NaN   
4                  3.5               28.0   2.99              NaN   

                                         Description  \
0  Join over 21,000,000 of our fans and download ...   
1  The classic game of Reversi, also known as Oth...   
2  Play the classic strategy game Othello (also k...   
3  Top 100 free app for over a year.\nRated "Best...   
4  "Senet Deluxe - The Ancient Game of Life and A...   

                  Developer Age Rating  \
0  Mighty Mighty Good Games         4+   
1          Kiss The Machine         4+   
2               Bayou Games         4+   
3  Mighty Mighty Good Games         4+   
4           RoGame Software         4+   

                                           Languages        Size  \
0  DA, NL, EN, FI, FR, DE, IT, JA, KO, NB, PL, PT...  15853568.0   
1                                                 EN  12328960.0   
2                                                 EN    674816.0   
3  DA, NL, EN, FI, FR, DE, IT, JA, KO, NB, PL, PT...  21552128.0   
4  DA, NL, EN, FR, DE, EL, IT, JA, KO, NO, PT, RU...  34689024.0   

  Primary Genre                             Genres Original Release Date  \
0         Games            Games, Strategy, Puzzle            11/07/2008   
1         Games             Games, Strategy, Board            11/07/2008   
2         Games             Games, Board, Strategy            11/07/2008   
3         Games            Games, Strategy, Puzzle            23/07/2008   
4         Games  Games, Strategy, Board, Education            18/07/2008   

  Current Version Release Date  
0                   30/05/2017  
1                   17/05/2018  
2                    5/09/2017  
3                   30/05/2017  
4                   22/07/2018

In [4]:
data = []
for i in games.columns:
    data.append([i, games[i].dtype, games[i].isna().sum(), 
                 round(games[i].isna().sum()*100/len(games[i]), 2), 
                 games[games[i].notna()][i].nunique(), 
                 games[games[i].notna()][i].drop_duplicates().sample(2).values])

pd.DataFrame(data, columns = ['dataFeatures', 'dataType', 'null', 'nullPct', 'unique', 'uniqueSample'])

dataFeatures dataType   null  nullPct  unique  \
0                            URL   object      0     0.00   16847   
1                             ID    int64      0     0.00   16847   
2                           Name   object      0     0.00   16847   
3                       Subtitle   object  11746    69.07    5010   
4                       Icon URL   object      0     0.00   16847   
5            Average User Rating  float64   9446    55.54       9   
6              User Rating Count  float64   9446    55.54    1792   
7                          Price  float64     24     0.14      24   
8               In-app Purchases   object   9324    54.82    3803   
9                    Description   object      0     0.00   16473   
10                     Developer   object      0     0.00    8693   
11                    Age Rating   object      0     0.00       4   
12                     Languages   object     60     0.35     990   
13                          Size  float64      1     0.01   15794   
14                 Primary Genre   object      0     0.00      21   
15                        Genres   object      0     0.00    1004   
16         Original Release Date   object      0     0.00    3084   
17  Current Version Release Date   object      0     0.00    2512   

                                         uniqueSample  
0   [https://apps.apple.com/us/app/2012-zombies-vs...  
1                             [828566277, 1356038933]  
2   [Take Off Master - Backyard Plane Simulator, S...  
3   [Innovative match3 game, Pick axe to rise arme...  
4   [https://is1-ssl.mzstatic.com/image/thumb/Purp...  
5                                          [2.0, 3.0]  
6                                      [515.0, 102.0]  
7                                        [7.99, 6.99]  
8   [0.99, 2.99, 1.99, 9.99, 4.99, 9.99, 4.99, 19....  
9   [Merge your knives and swords to make them mor...  
10             [Twisted Games, Tigers Agreement GmbH]  
11                                          [17+, 9+]  
12  [AR, CA, HR, CS, DA, NL, EN, FI, FR, DE, EL, H...  
13                           [20501504.0, 14944256.0]  
14                               [Education, Medical]  
15  [Games, Trivia, Productivity, Strategy, News, ...  
16                           [24/05/2011, 17/07/2012]  
17                             [2/02/2012, 6/10/2018]

## Data Cleaning & Analysis (without using visualization)

In [5]:
games_2 = games.copy()

In [6]:
games.groupby(['Primary Genre']).mean()['Average User Rating']

Primary Genre
Book                 4.300000
Business             3.000000
Education            4.152174
Entertainment        3.831522
Finance              4.062500
Food & Drink         5.000000
Games                4.065903
Health & Fitness     4.500000
Lifestyle            3.125000
Medical              3.000000
Music                4.500000
Navigation           4.000000
News                 4.750000
Productivity         3.250000
Reference            3.361111
Shopping             4.500000
Social Networking    3.833333
Sports               4.217391
Stickers             3.666667
Travel               3.500000
Utilities            4.056818
Name: Average User Rating, dtype: float64

In [7]:
mean_all = games.groupby(['Primary Genre']).mean()
mean_12 = games[games['Age Rating'] == '12+'].groupby(['Primary Genre']).mean()

def fill_na(df):
    if df['Age Rating'] == '4+' or df['Age Rating'] == '17+':
        return 5.0
    elif df['Age Rating'] == '9+':
        return mean_all.loc[df['Primary Genre'], 'Average User Rating']
    else:
        return mean_12.loc[df['Primary Genre'], 'Average User Rating']
    
games_2['Average User Rating'].fillna(value = games_2.apply(fill_na, axis = 1), inplace = True)

In [8]:
games_2['Average User Rating'].value_counts()

5.000000    8705
4.500000    2868
4.000000    1727
4.065903     984
3.500000     925
4.123648     699
3.000000     515
2.500000     317
2.000000     158
1.500000      60
1.000000      14
3.831522      11
3.607143       8
4.152174       4
3.361111       3
4.750000       3
4.056818       3
4.300000       2
3.666667       1
Name: Average User Rating, dtype: int64

In [9]:
# Looking for data duplicates
len(games[games.duplicated(keep = 'last')])

160

In [10]:
# Drop all duplicate datas but keep the last data
games.drop_duplicates(keep = 'last', inplace = True)
games.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16847 entries, 0 to 17006
Data columns (total 18 columns):
URL                             16847 non-null object
ID                              16847 non-null int64
Name                            16847 non-null object
Subtitle                        5212 non-null object
Icon URL                        16847 non-null object
Average User Rating             7488 non-null float64
User Rating Count               7488 non-null float64
Price                           16823 non-null float64
In-app Purchases                7615 non-null object
Description                     16847 non-null object
Developer                       16847 non-null object
Age Rating                      16847 non-null object
Languages                       16787 non-null object
Size                            16846 non-null float64
Primary Genre                   16847 non-null object
Genres                          16847 non-null object
Original Release Date       

Hubungan jumlah kata dengan harga

Berdasarkan analisis yang dilakukan, saya berkesimpulan bahwa Jumlah Kata dari Feature Description secara umum **tidak** mempengaruhi Feature Price. Hal ini terlihat dari rata-rata jumlah kata untuk masing-masing Price. Pada beberapa rentang harga (misalnya antara 0.00 - 6.99 dan 29.99 - 59.99), rata-rata Jumlah Kata dari Description cenderung naik seiring dengan semakin tingginya Price. Akan tetapi secara keseluruhan, Jumlah Kata dari Description tidak berbanding lurus (cenderung fluktuatif) terhadap Price. 

In [11]:
# Cleaning data of Feature Description & making Feature n_word (amount of words)
import re

def count_words(desc):
    new_desc = re.sub(r'(\\[a-z])+(\d+)*| - | [*] ', ' ', desc)
    return len(re.findall(r'\S+', desc))

games['n_word'] = games['Description'].apply(count_words)

In [12]:
# Amount of words in Description for games with Price of 0.00
games[(games['Price'].notna()) & (games['Price'] == games['Price'].min())][['Price', 'n_word']]

Price  n_word
2        0.0      97
3        0.0     267
5        0.0     368
6        0.0     109
8        0.0      61
9        0.0      83
10       0.0     110
11       0.0     101
12       0.0     104
14       0.0     163
16       0.0     101
17       0.0      97
18       0.0     110
19       0.0     170
22       0.0      98
25       0.0     270
27       0.0     135
31       0.0     122
37       0.0     152
38       0.0     142
40       0.0     214
48       0.0      99
50       0.0     181
51       0.0     113
54       0.0     378
58       0.0     242
60       0.0     374
61       0.0     334
62       0.0     129
63       0.0     340
...      ...     ...
16970    0.0      49
16971    0.0      42
16972    0.0      78
16973    0.0      11
16974    0.0      37
16975    0.0      16
16976    0.0      16
16977    0.0      82
16979    0.0      31
16981    0.0      21
16982    0.0      28
16983    0.0      22
16984    0.0      38
16985    0.0     339
16986    0.0      36
16988    0.0     100
16989    0.0      10
16990    0.0      56
16991    0.0      83
16992    0.0      31
16993    0.0      45
16996    0.0      23
16997    0.0      52
16998    0.0      37
16999    0.0      68
17002    0.0      84
17003    0.0      27
17004    0.0      70
17005    0.0      24
17006    0.0     114

[14084 rows x 2 columns]

In [13]:
# Amount of words in Description for games with Price of 179.99
games[(games['Price'].notna()) & (games['Price'] == games['Price'].max())][['Price', 'n_word']]

Price  n_word
8177   179.99     258
8196   179.99     270
8198   179.99     314
8200   179.99     296
8273   179.99     377
8281   179.99     328
8292   179.99     360
8301   179.99     387
8303   179.99     366
8372   179.99     364
8548   179.99     307
8600   179.99     338
8771   179.99     366
8807   179.99     391
8887   179.99     359
9005   179.99     324
9023   179.99     364
9025   179.99     282
9093   179.99     284
9159   179.99     315
9219   179.99     314
9276   179.99     309
9308   179.99     277
9409   179.99     258
9411   179.99     301
9491   179.99     324
9545   179.99     346
9550   179.99     272
9724   179.99     287
12437  179.99     297

In [14]:
# Average of amount of words in Description for games grouped by Price
games.groupby('Price').mean()['n_word']

Price
0.00      153.098978
0.99      142.946479
1.99      168.150391
2.99      203.229614
3.99      255.856250
4.99      267.659483
5.99      260.166667
6.99      343.806452
7.99      326.176471
8.99      184.307692
9.99      243.290698
11.99      74.355932
12.99     283.000000
14.99     378.500000
16.99     240.000000
18.99     403.000000
19.99     297.111111
29.99     142.000000
36.99     188.000000
37.99     362.000000
59.99     422.000000
99.99      97.000000
139.99    457.000000
179.99    321.166667
Name: n_word, dtype: float64

Null value pada Average User Rating, User Rating Count, dan In-app Purchase

   Feature Average User Rating sangat tergantung pada Feature User Rating Count. Jika tidak ada user yang memberikan rating maka tidak mungkin terdapat nilai rata-rata (average) dari rating tersebut. Hal ini terlihat dimana Null Value pada kedua Feature tersebut memiliki jumlah yang sama. Jika ditarik data dari User Rating Count yang memiliki Null Value, maka semua data Average User Rating juga memiliki Null Value.
   
   Kesimpulannya, banyaknya Null Value pada Feature Average User Rating dikarenakan banyak juga Null Value pada Feature User Rating Count (penyebab Null pada User Rating Count dijelaskan di bawah ini.) 

In [15]:
# All Average User Rating datas only contain Null Value when User Rating Count is Null
games[games['User Rating Count'].isna()]['Average User Rating'].unique()

array([nan])

   Terkait banyaknya Null Value Feature User Rating Count, saya tidak melihat adanya hubungan yang kuat antara Feature ini dengan Feature lainnya. Saya memiliki asumsi bahwa terdapat hubungan antara Feature User Rating Count dengan Feature Age Rating dimana games dengan rating umur 4+ cenderung tidak memiliki rating (dengan asumsi bahwa anak kecil tidak terlalu memikirkan untuk memberikan rating) akan tetapi perbedaan yang ada kurang signifikan yaitu terdapat 75% games tanpa rating user dengan rating umur 4+ dibandingkan dengan 60% games dengan rating user pada rating umur yang sama. 
   
   Saya mencoba mencari mekanisme terkait pemberian rating di app store dan saya menemukan informasi terkait hal ini dikutip dari laman web https://developer.apple.com/app-store/ratings-and-reviews/: _"This summary rating is specific to each territory on the App Store and you can reset it when you release a new version of your app."_ Hal ini berarti terdapat kemungkinan bahwa rating user hanya dapat ditampilkan pada daerah tertentu dan kemungkinan lainnya adalah pengembang me-reset rating user pada saat mengeluarkan versi terbarunya. Saya berasumsi hal ini dilakukan pengembang karena aplikasi tersebut mendapatkan rating yang kurang baik. Saya juga mencoba melihat distribusi dari rating user dan memang sangat sedikit games yang memiliki rating dibawah 2.5 (kemungkinan memang lebih banyak games yang bagus atau kemungkinan pengembang me-reset rating jika games tersebut memiliki rating yang kurang baik).
   
   Kesimpulannya, saya berasumsi bahwa banyaknya Null Value pada Feature User Rating Count umumnya disebabkan oleh pengembang yang me-reset rating games tersebut setiap mereka merilis versi terbaru. Hal ini dilakukan dengan asumsi bahwa games tersebut memiliki rating yang kurang baik dan berakibat user lainnya enggan mengunduh games tersebut akibat tidak adanya rating yang menunjukkan kualitas games tersebut. Hal ini juga berakibat terhadap banyaknya Null Value pada Feature Average User Rating.  

In [16]:
# Amount of games without Average User Rating grouped by Age Rating
games[games['User Rating Count'].isna()]['Age Rating'].value_counts()

4+     7269
9+     1003
12+     713
17+     374
Name: Age Rating, dtype: int64

In [17]:
# Amount of games with Average User Rating grouped by Age Rating
games[games['User Rating Count'].notna()]['Age Rating'].value_counts()

4+     4424
9+     1456
12+    1321
17+     287
Name: Age Rating, dtype: int64

In [18]:
# Amount of games grouped by Average User Rating
games['Average User Rating'].value_counts()

4.5    2832
4.0    1707
5.0     984
3.5     916
3.0     507
2.5     313
2.0     157
1.5      59
1.0      13
Name: Average User Rating, dtype: int64

   Terkait banyaknya Null Value pada Feature In-app Purchases, saya berasumsi bahwa Feature ini tergantung dari harga games itu sendiri (Feature Price). Asumsinya bahwa semakin mahal harga suatu game, maka umumnya game tersebut tidak memiliki in-app purchase. Berdasarkan data games yang ada, tidak terdapat perbedaan yang signifikan terutama untuk games dibawah 20.00. Jumlah games gratis (seharga 0.00) yang memiliki maupun tidak memiliki in-app purchase hampir sama (sekitar 7000-an). Games dengan rentang harga antara 0.99 dan 4.99 umumnya tidak memiliki in-app purchase (1965 vs 499). Semua games dengan harga diatas 20.00 sama sekali tidak memiliki in-app purchase (kecuali 1 games dengan harga 139.99).
   
   Kesimpulannya, saya berasumsi bahwa semakin mahal harga suatu games, maka games tersebut biasanya tidak akan memiliki In-app Purchases. Akan tetapi, pada rentang harga dibawah 20.00 tidak terdapat perbedaan yang signifikan antara games yang memiliki dan tidak memiliki In-app Purchases. Saya juga berasumsi ada faktor lain yang mempengaruhi ada tidaknya In-app Purchases pada suatu games misalnya jenis games (casual, RPG, dan sebagainya).

In [19]:
# Amount of games without In-app Purchases
games['In-app Purchases'].isna().sum()

9232

In [20]:
# Amount of games without In-app Purchases grouped by Price
games[games['In-app Purchases'].isna()]['Price'].value_counts()

0.00      7055
0.99       877
1.99       410
2.99       360
4.99       166
3.99       130
9.99        75
5.99        32
179.99      30
6.99        26
11.99       17
7.99        15
8.99        10
19.99        5
14.99        3
16.99        2
29.99        2
59.99        1
36.99        1
99.99        1
12.99        1
37.99        1
Name: Price, dtype: int64

In [21]:
# Amount of games with In-app Purchases grouped by Price
games[games['In-app Purchases'].notna()]['Price'].value_counts()

0.00      7029
0.99       188
2.99       106
1.99       102
4.99        66
11.99       42
3.99        30
9.99        11
5.99        10
6.99         5
19.99        4
8.99         3
12.99        2
7.99         2
14.99        1
139.99       1
18.99        1
Name: Price, dtype: int64

Hubungan antara harga dengan Original Release Date dan Current Version Release Date

   Berdasarkan analisis yang dilakukan, saya berkesimpulan bahwa Feature Original Release Date dan Current Value Release Date **tidak** mempengaruhi Feature Price. Berdasarkan distribusi dari data Price maka terlihat bahwa sekitar 85% data tersebut terpusat pada harga senilai 0.00. Saya mencoba melihat penyebaran berdasarkan tahun pada Feature Original Release Date dan Current Value Release Date pada nilai harga tertentu.
   
   Berdasarkan Feature Original Release Date, diketahui bahwa sekitar 50% games yang berada di App Store dirilis antara tahun 2016 - 2018. Games seharga dibawah 5.00 umumnya (sekitar 50%) dirilis sekitar tahun 2016 - 2018 sedangkan semua games dengan harga tertinggi (179.99) dirilis tahun 2016 (kecuali satu game yang dirilis tahun 2017). Saya juga mencoba melihat penyebaran harga dari games yang dirilis pada 5 tahun pertama (2008 - 2012) dan memang harga paling dominan adalah yang senilai 0.00 (mengingat distribusi Price yang sangat terpusat pada nilai tersebut). Oleh karena itu, kami tidak melihat adanya pengaruh antara Feature Original Release Date dan Feature Price.
   
   Berdasarkan Feature Current Value Release Date, diketahui bahwa sekitar 60% games yang berada di App Store dirilis selama 3 tahun terakhir (2017 - 2019). Games seharga dibawah 5.00 umumnya (sekitar 60%) terakhir diperbaharui sekitar tahun 2017 - 2019 sedangkan games dengan harga tertinggi (179.99) terakhir diperbaharui pada tahun 2016 dan 2017. Penyebaran harga dari games yang terakhir diperbaharui pada 5 tahun pertama (2008 - 2012) juga paling dominan pada harga senilai 0.00. Oleh karena itu, kami tidak melihat adanya pengaruh antara Feature Current Version Release Date dan Feature Price.

In [22]:
# Making new Feature named as Original Year and Current Year
games['Original Release Date'] = games['Original Release Date'].astype('datetime64')
games['Current Version Release Date'] = games['Current Version Release Date'].astype('datetime64')
games['Current Year'] = games['Current Version Release Date'].apply(lambda x: x.year)
games['Original Year'] = games['Original Release Date'].apply(lambda x: x.year) 

In [23]:
games['Price'].value_counts()

0.00      14084
0.99       1065
1.99        512
2.99        466
4.99        232
3.99        160
9.99         86
11.99        59
5.99         42
6.99         31
179.99       30
7.99         17
8.99         13
19.99         9
14.99         4
12.99         3
16.99         2
29.99         2
99.99         1
36.99         1
59.99         1
139.99        1
37.99         1
18.99         1
Name: Price, dtype: int64

In [24]:
games['Original Year'].value_counts()

2016    3099
2017    2793
2018    2484
2015    2190
2014    1876
2019    1552
2013    1150
2012     723
2011     483
2010     276
2009     175
2008      46
Name: Original Year, dtype: int64

In [25]:
# Amount of games with Price less than 5.00 grouped by Original Year of release
games[games['Price'] <= 5.00]['Original Year'].value_counts()

2016    3006
2017    2726
2018    2460
2015    2169
2014    1854
2019    1506
2013    1128
2012     713
2011     471
2010     270
2009     173
2008      43
Name: Original Year, dtype: int64

In [26]:
# Amount of games with highest price (179.99) grouped by Original Year of release
games[games['Price'] == max(games['Price'])]['Original Year'].value_counts()

2016    29
2017     1
Name: Original Year, dtype: int64

In [27]:
# Amount of games originally released between 2008 - 2012 grouped by Price
games[games['Original Year'] <= 2012]['Price'].value_counts()

0.00     1161
0.99      236
2.99       99
1.99       94
4.99       50
3.99       30
5.99        9
9.99        9
7.99        7
6.99        4
11.99       2
8.99        1
19.99       1
Name: Price, dtype: int64

In [28]:
games['Current Year'].value_counts()

2019    3744
2017    3459
2018    3182
2016    2854
2015    1608
2014    1088
2013     510
2012     224
2011     108
2010      47
2009      17
2008       6
Name: Current Year, dtype: int64

In [29]:
# Amount of games with Price less than 5.00 grouped by Current Year (latest year of update)
games[games['Price'] <= 5.00]['Current Year'].value_counts()

2019    3662
2017    3343
2018    3129
2016    2799
2015    1592
2014    1087
2013     507
2012     223
2011     107
2010      47
2009      17
2008       6
Name: Current Year, dtype: int64

In [30]:
# Amount of games with highest price (179.99) grouped by Current Year (latest year of update)
games[games['Price'] == max(games['Price'])]['Current Year'].value_counts()

2016    16
2017    14
Name: Current Year, dtype: int64

In [31]:
# Amount of games which is last updated between 2008 - 2012 grouped by Price
games[games['Current Year'] <= 2012]['Price'].value_counts()

0.00    368
0.99     19
1.99      8
2.99      5
9.99      1
5.99      1
Name: Price, dtype: int64